File: files/2025-08-24/CSEC 24082025.py
import pandas as pd

path = "files/2025-08-24/Challenge 52.xlsx"

input = pd.read_excel(path, usecols="B:E", skiprows=1, nrows=6, dtype = str)
test = pd.read_excel(path, usecols="G:H", skiprows=1, nrows=6).rename(columns=lambda c: c.replace('.1', '')).assign(**{"Missing Data": lambda df: df["Missing Data"].fillna("")})

input["Customer"] = input["Customer"].astype("category")

long = input.melt(id_vars="Customer", var_name="Var", value_name="val")
long["val"] = long["val"].astype(str)
long.loc[long["val"] == "nan", "val"] = None
long["missing"] = long.apply(lambda row: row["Var"] if pd.isna(row["val"]) else None, axis=1)

result = (
    long.groupby("Customer")["missing"]
    .apply(lambda x: ", ".join([v for v in x if pd.notna(v)]))
    .reset_index()
    .rename(columns={"missing": "Missing Data"})
)

result["Customer"] = result["Customer"].astype(str)

print(result.equals(test))

File: files/2025-08-24/CSEC 24082025.R
library(tidyverse)
library(readxl)

path = "files/2025-08-24/Challenge 52.xlsx"
input = read_excel(path, range = "B2:E8")
test  = read_excel(path, range = "G2:H8") %>%
  replace_na(list(`Missing Data` = ""))

result = input %>%
  mutate(Customer = as.factor(Customer)) %>%
  pivot_longer(cols = -Customer, 
               names_to = "Var", 
               values_to = "val", 
               values_transform = as.character) %>%
  mutate(missing = ifelse(is.na(val), Var, NA)) %>%
  summarise(`Missing Data` = paste(missing[!is.na(missing)], collapse = ", "), .by = Customer) %>%
  mutate(Customer = as.character(Customer))

all.equal(result, test, check.attributes = FALSE)
# > [1] TRUE         